### PR TITLE
Fix mark and unmark commands

### PR DESCRIPTION
**Pull request by Gordon25**
- [x] fix bug in mark/unmark parser (unable to parser index > 9) 
- [x] fix order at which contacts / events are marked
- [x] fix bugs in cunmark, enumark command test
- [x] make mark/unmark compatible with undo/redo
- [x] make esort compatible with mark events
- [x] fix bug with multiple marking of a single contact/ event
- [x] fix bug in mark and unmark parser that allows empty argument

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/contact/CUnmarkCommand.java [hunk 1]
```diff
@@ -26,7 +27,7 @@
             + "Example: " + COMMAND_WORD + " 1";
 
     public static final String MESSAGE_SUCCESS = "Unmarked Contact: %1$s";
-    public static final String MESSAGE_NOT_MARKED = "Contact %1$s is not bookmarked!";
+    public static final String MESSAGE_NOT_MARKED = "Contact %1$s is not mrked!";
```
Comment (chunweii) [inline src/main/java/seedu/address/logic/commands/contact/CUnmarkCommand.java:30]: marked

File: src/main/java/seedu/address/logic/commands/event/EMarkCommand.java [hunk 2]
```diff
@@ -25,15 +27,15 @@
             + "Parameters: " + PARAMETERS
             + "Example: " + COMMAND_WORD + " 1";
 
-    public static final String MESSAGE_SUCCESS = "Bookmarked Event: %1$s";
-    public static final String MESSAGE_ALREADY_MARKED = "Event %1$s is already bookmarked!";
+    public static final String MESSAGE_SUCCESS = "Marked Event: %1$s";
+    public static final String MESSAGE_ALREADY_MARKED = "Event Already Marked: %1$s";
```
Comment (chunweii) [inline src/main/java/seedu/address/logic/commands/event/EMarkCommand.java:31]: small letters

File: src/main/java/seedu/address/logic/commands/event/EUnmarkCommand.java [hunk 3]
```diff
@@ -26,7 +27,7 @@
             + "Example: " + COMMAND_WORD + " 1";
 
     public static final String MESSAGE_SUCCESS = "Unmarked Event: %1$s";
-    public static final String MESSAGE_NOT_MARKED = "Event %1$s is not bookmarked!";
+    public static final String MESSAGE_NOT_MARKED = "Event %1$s is not mrked!";
```
Comment (chunweii) [inline src/main/java/seedu/address/logic/commands/event/EUnmarkCommand.java:30]: marked

File: src/main/java/seedu/address/logic/parser/event/EMarkCommandParser.java [hunk 4]
```diff
@@ -5,24 +5,24 @@
...
-import seedu.address.logic.commands.event.EBookmarkCommand;
+import seedu.address.logic.commands.event.EMarkCommand;
 import seedu.address.logic.parser.Parser;
 import seedu.address.logic.parser.ParserUtil;
 import seedu.address.logic.parser.exceptions.ParseException;
 
-public class EBookmarkCommandParser implements Parser<EBookmarkCommand> {
+public class EMarkCommandParser implements Parser<EMarkCommand> {
     /**
      * Parses the given {@code String} of arguments in the context of the EBookmarkCommand
```
Comment (chunweii) [inline src/main/java/seedu/address/logic/parser/event/EMarkCommandParser.java:15]: EMarkCommand

File: src/main/java/seedu/address/logic/parser/contact/CMarkCommandParser.java [hunk 5]
```diff
@@ -5,24 +5,24 @@
...
-import seedu.address.logic.commands.contact.CBookmarkCommand;
+import seedu.address.logic.commands.contact.CMarkCommand;
 import seedu.address.logic.parser.Parser;
 import seedu.address.logic.parser.ParserUtil;
 import seedu.address.logic.parser.exceptions.ParseException;
 
-public class CBookmarkCommandParser implements Parser<CBookmarkCommand> {
+public class CMarkCommandParser implements Parser<CMarkCommand> {
     /**
      * Parses the given {@code String} of arguments in the context of the CBookmarkCommand
```
Comment (chunweii) [inline src/main/java/seedu/address/logic/parser/contact/CMarkCommandParser.java:15]: CMarkCommand

File: src/main/java/seedu/address/model/AddressBook.java [hunk 6]
```diff
@@ -90,10 +91,12 @@ public void resetData(ReadOnlyAddressBook newData) {
     }
 
     /**
-     * Places bookmarked {@code contacts} at the top of the list.
+     * Places marked {@code contacts} at the top of the list.
+     * Places the newly marked contacts or replaces newly unmarked contacts
+     * in the order specified in {code indexes} if specified.
```
Comment (chunweii) [inline src/main/java/seedu/address/model/AddressBook.java:96]: {@code indexes}

File: src/main/java/seedu/address/model/AddressBook.java [hunk 7]
```diff
@@ -187,10 +190,12 @@ public void resetEvents() {
     }
 
     /**
-     * Places bookmarked {@code events} at the top of the list.
+     * Places marked {@code events} at the top of the list.
+     * Places the newly marked events or replaces newly unmarked events
+     * in the order specified in {code indexes} if specified.
```
Comment (chunweii) [inline src/main/java/seedu/address/model/AddressBook.java:195]: {@code indexes}

File: src/main/java/seedu/address/model/contact/Contact.java [hunk 8]
```diff
@@ -36,7 +36,7 @@
     private final ZoomLink zoomLink;
     private final Set<Tag> tags = new HashSet<>();
     private final Set<UUID> linkedEvents = new HashSet<>();
-    private boolean isBookMarked;
+    private boolean isMarked;
```
Comment (chunweii) [inline src/main/java/seedu/address/model/contact/Contact.java:39]: Why is this not set to final?

File: src/main/java/seedu/address/model/contact/Contact.java [hunk 9]
```diff
@@ -123,12 +122,12 @@ public UUID getUuid() {
...
 
-    public boolean getIsBookMarked() {
-        return isBookMarked;
+    public boolean getIsMarked() {
+        return isMarked;
     }
 
-    public void setBookMarked(boolean bookMarked) {
-        isBookMarked = bookMarked;
+    public void setMarked(boolean marked) {
```
Comment (chunweii) [inline src/main/java/seedu/address/model/contact/Contact.java:129]: Remove this, make contact immutable.

File: src/main/java/seedu/address/model/contact/UniqueContactList.java [hunk 10]
```diff
@@ -107,22 +111,35 @@ public void resetContacts() {
...
+                    .forEach(contact -> tempList.add(contact));
+            internalList.removeAll(internalUnmodifiableList);
+            internalList.addAll(tempList);
+        }
+    }
+
+    private void transferElementsToNewList(Predicate<Contact> predicate, ObservableList<Contact> originalList,
+                                           Deque<Contact> newList) {
+        originalList.filtered(predicate).forEach(contact -> newList.add(contact));
+        originalList.filtered(predicate).forEach(contact -> originalList.remove(contact));
```
Comment (chunweii) [inline src/main/java/seedu/address/model/contact/UniqueContactList.java:142]: What is this for?

File: src/main/java/seedu/address/model/contact/UniqueContactList.java [hunk 11]
```diff
@@ -107,22 +111,35 @@ public void resetContacts() {
     }
 
     /**
-     * Moves bookmarked contacts to the top of the list.
+     * Moves marked contacts to the top of the list.
+     * Places the newly marked contacts or replaces newly unmarked contacts
+     * in the order specified in {@code indexes} if specified.
```
Comment (chunweii) [inline src/main/java/seedu/address/model/contact/UniqueContactList.java:116]: What is `isMark` for? Can you mention this in the javadoc?

File: src/main/java/seedu/address/model/event/Event.java [hunk 12]
```diff
@@ -37,7 +37,7 @@
 
     private final Set<Tag> tags = new HashSet<>();
     private final Set<UUID> linkedContacts = new HashSet<>();
-    private boolean isBookMarked;
+    private boolean isMarked;
```
Comment (chunweii) [inline src/main/java/seedu/address/model/event/Event.java:40]: Should be final

File: src/main/java/seedu/address/model/event/Event.java [hunk 13]
```diff
@@ -124,13 +124,14 @@ public UUID getUuid() {
...
-    public boolean getIsBookMarked() {
-        return isBookMarked;
+    public boolean getIsMarked() {
+        return isMarked;
     }
 
-    public void setBookMarked(boolean bookMarked) {
-        isBookMarked = bookMarked;
+    public void setMarked(boolean isMarked) {
+        this.isMarked = isMarked;
```
Comment (chunweii) [inline src/main/java/seedu/address/model/event/Event.java:132]: Remove this method. Event should be immutable.

File: src/main/java/seedu/address/model/event/UniqueEventList.java [hunk 14]
```diff
@@ -107,43 +104,45 @@ public void setEvents(List<Event> events) {
...
-    public void bookmarkEvent(Index index) {
-        Event eventToMark = internalUnmodifiableList.get(index.getZeroBased());
-        eventToMark.setBookMarked(true);
-    }
-
-    /**
-     * Moves bookmarked events to the top of the list.
+     * Moves marked events to the top of the list.
+     * Places the newly marked events or replaces newly unmarked events
+     * in the order specified in {@code indexes} if specified.
```
Comment (chunweii) [inline src/main/java/seedu/address/model/event/UniqueEventList.java:125]: Explain what `isMarked` is

**Review by chunweii — CHANGES_REQUESTED** [review]
LGTM, except for a few issues:

1. Remove the UG changes, since we are going to change it in the ug branch.
2. There is a bug which will still mark an event/contact if there is an invalid index. `cmark 100 1` will mark 1.
3. If I enter `emark` or `eunmark` or `cmark` or `cunmark` by itself, nothing happens.
4. See other comments

File: src/main/java/seedu/address/model/contact/Contact.java (hunk 8 above)
Comment (Gordon25) [reply src/main/java/seedu/address/model/contact/Contact.java:39]: If `isMarked` is set to final, then i cannot mark or unmark a contact using `setMarked`. Cannot change values of a final variable

Comment (Gordon25) [reply src/main/java/seedu/address/model/contact/Contact.java:39]: I need `setMarked` for creating contacts in `ContactBuilder`, i will create a new constructor just for `ContactBuilder` and make `isMark` final

File: src/main/java/seedu/address/logic/commands/contact/CMarkCommand.java [hunk 15]
```diff
@@ -27,7 +27,7 @@
             + "Example: " + COMMAND_WORD + " 1";
 
     public static final String MESSAGE_SUCCESS = "Marked Contact: %1$s";
-    public static final String MESSAGE_ALREADY_MARKED = "Contact Already Marked: %1$s";
+    public static final String MESSAGE_ALREADY_MARKED = "Contact already Marked: %1$s";
```
Comment (chunweii) [inline src/main/java/seedu/address/logic/commands/contact/CMarkCommand.java:30]: You did not fix this completely. Marked should be in small letters

File: src/main/java/seedu/address/logic/commands/event/EMarkCommand.java [hunk 16]
```diff
@@ -28,7 +28,7 @@
             + "Example: " + COMMAND_WORD + " 1";
 
     public static final String MESSAGE_SUCCESS = "Marked Event: %1$s";
-    public static final String MESSAGE_ALREADY_MARKED = "Event Already Marked: %1$s";
+    public static final String MESSAGE_ALREADY_MARKED = "Event already Marked: %1$s";
```
Comment (chunweii) [inline src/main/java/seedu/address/logic/commands/event/EMarkCommand.java:31]: Same. Put small letters

**Review by chunweii — CHANGES_REQUESTED** [review]
Some typo errors. There is another bug (See previous review point 3):
1. If I enter emark or eunmark or cmark or cunmark by itself, nothing happens.

**Review by chunweii — APPROVED** [review]
LGTM

**Pull request merged**